flang/include/flang/Optimizer/Dialect/FIROps.td

```diff
@@ -1133,15 +1134,11 @@ def fir_EmboxOp : fir_Op<"embox", [NoSideEffect, AttrSizedOperandSegments]> {
         if (lenPs != rt.getNumLenParams())
           return emitOpError("number of LEN params does not correspond"
                              " to the !fir.type type");
+      } else if (auto strTy = fir::unwrap_char(eleTy)) {
```

Comment:
Since `unwrap_char` dives through box types, it seems suspect here. Boxing a box type should be invalid.

Reply:
Right, this is unsafe. I modified this code to peel potential the sequence type before this and to use a `dyn_cast<fir::CharacterType>()` here. I also took the opportunity of peeling the sequence to add a check that shape/slice are not provided for scalars.